src/consensus/encode.rs

```diff
@@ -34,21 +34,327 @@ use std::hash::Hash;
+            Error::UnexpectedNetworkMagic { .. } => "unexpected network magic",
+            Error::OversizedVectorAllocation { .. } => "allocation of oversized vector requested",
+            Error::InvalidChecksum { .. } => "invalid checksum",
+            Error::UnknownNetworkMagic(..) => "unknown network magic",
+            Error::ParseFailed(..) => "parse failed",
+            Error::UnsupportedWitnessVersion(..) => "unsupported witness version",
+            Error::UnsupportedSegwitFlag(..) => "unsupported segwit version",
+            Error::UnrecognizedNetworkCommand(..) => "unrecognized network command",
+            Error::UnexpectedHexDigit(..) => "unexpected hex digit",
+        }
+    }
+}
+
+#[doc(hidden)]
+impl From<base58::Error> for Error {
+    fn from(e: base58::Error) -> Error {
+        Error::Base58(e)
+    }
+}
+
+#[doc(hidden)]
+impl From<bitcoin_bech32::Error> for Error {
+    fn from(e: bitcoin_bech32::Error) -> Error {
+        Error::Bech32(e)
+    }
+}
+
+
+#[doc(hidden)]
+impl From<io::Error> for Error {
+    fn from(error: io::Error) -> Self {
+        Error::Io(error)
+    }
+}
+
+/// Encode an object into a vector
+pub fn serialize<T: ?Sized>(data: &T) -> Result<Vec<u8>, Error>
+     where T: Encodable<RawEncoder<Cursor<Vec<u8>>>>,
+{
+    let mut encoder = RawEncoder::new(Cursor::new(vec![]));
+    data.consensus_encode(&mut encoder)?;
+    Ok(encoder.into_inner().into_inner())
+}
+
+/// Encode an object into a hex-encoded string
+pub fn serialize_hex<T: ?Sized>(data: &T) -> Result<String, Error>
+     where T: Encodable<RawEncoder<Cursor<Vec<u8>>>>
+{
+    let serial = serialize(data)?;
+    Ok(hex_encode(serial))
+}
+
+/// Deserialize an object from a vector, will error if said deserialization
+/// doesn't consume the entire vector.
+pub fn deserialize<'a, T>(data: &'a [u8]) -> Result<T, Error>
+     where T: Decodable<RawDecoder<Cursor<&'a [u8]>>>
+{
+    let mut decoder = RawDecoder::new(Cursor::new(data));
+    let rv = Decodable::consensus_decode(&mut decoder)?;
+
+    // Fail if data is not consumed entirely.
+    if decoder.into_inner().position() == data.len() as u64 {
+        Ok(rv)
+    } else {
+        Err(Error::ParseFailed("data not consumed entirely when explicitly deserializing"))
+    }
+}
+
+/// An encoder for raw binary data
+pub struct RawEncoder<W> {
+    writer: W
+}
+
+/// An decoder for raw binary data
+pub struct RawDecoder<R> {
+    reader: R
+}
+
+impl<W: Write> RawEncoder<W> {
+    /// Constructor
+    pub fn new(writer: W) -> RawEncoder<W> { RawEncoder { writer: writer } }
+    /// Returns the underlying Writer
+    pub fn into_inner(self) -> W { self.writer }
+}
+
+impl<R: Read> RawDecoder<R> {
+  /// Constructor
+  pub fn new(reader: R) -> RawDecoder<R> { RawDecoder { reader: reader } }
+  /// Returns the underlying Reader
+  pub fn into_inner(self) -> R { self.reader }
+}
+
+/// A simple Encoder trait
+pub trait SimpleEncoder {
```

Comment:
Should be renamed to `Encoder`. Ditto for `SimpleDecoder`.

Reply:
Addressed in: 58b66001fdfa538dae9448c05ac1f307a8225d97